src/backbone_router/backbone_agent.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+ *    Copyright (c) 2020, The OpenThread Authors.
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   The file implements the Thread Backbone agent.
+ */
+
+#include "backbone_router/backbone_agent.hpp"
+
+#if OTBR_ENABLE_BACKBONE_ROUTER
+
+#include <openthread/backbone_router_ftd.h>
+
+#include "common/code_utils.hpp"
+
+namespace otbr {
+namespace BackboneRouter {
+
+BackboneAgent::BackboneAgent(otbr::Ncp::ControllerOpenThread &aThread)
+    : mThread(aThread)
+    , mBackboneRouterState(OT_BACKBONE_ROUTER_STATE_DISABLED)
+{
+}
+
+void BackboneAgent::Init(const std::string &aThreadIfName, const std::string &aBackboneIfName)
+{
+    mSMCRouteManager.Init(aThreadIfName, aBackboneIfName);
+
+    HandleBackboneRouterState();
+}
+
+void BackboneAgent::HandleBackboneRouterState(void)
+{
+    otBackboneRouterState state      = otBackboneRouterGetState(mThread.GetInstance());
+    bool                  wasPrimary = (mBackboneRouterState == OT_BACKBONE_ROUTER_STATE_PRIMARY);
+
+    otbrLog(OTBR_LOG_DEBUG, "BackboneAgent: HandleBackboneRouterState: state=%d, mBackboneRouterState=%d", state,
+            mBackboneRouterState);
+    VerifyOrExit(mBackboneRouterState != state);
+
+    mBackboneRouterState = state;
+
+    if (!wasPrimary && IsPrimary())
+    {
+        OnBecomePrimary();
+    }
+    else if (wasPrimary && !IsPrimary())
+    {
+        OnResignPrimary();
+    }
```

Comment:
There is `VerifyOrExit(mBackboneRouterState != state);` above, which grantees there is only one true for `wasPrimary` and `IsPrimary()`

```suggestion
    if (IsPrimary())
    {
        OnBecomePrimary();
    }
    else if (wasPrimary)
    {
        OnResignPrimary();
    }
```

Reply:
It's true for `OnBecomePrimary`. But not true for `OnResignPrimary` because device can change state from Disabled to Secondary or Secondary to Disabled.

Reply:
I have simplified the code.